### PR TITLE
WIP: HBitcoin integration

### PR DESCRIPTION
**Pull request by nopara73**
Please review those you don't agree with.  
Do not merge until issue https://github.com/stratisproject/Breeze/issues/20 is resolved.

### CODE REVIEW

File: Breeze/src/Breeze.Wallet/Wrappers/WalletWrapper.cs [hunk 1]
```diff
@@ -71,14 +87,14 @@ public WalletModel Recover(string password, string folderPath, string name, stri
...
-			switch (network.ToLowerInvariant())
-			{
-				case "main":
-				case "mainnet":
-					return Network.Main;					
-				default:
-					return Network.TestNet;
-			}
+			var trimmed = network.Trim();
+			if (trimmed.Equals("main", StringComparison.OrdinalIgnoreCase)
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Wrappers/WalletWrapper.cs]: can do network.Trim().ToLowerInvariant() and use the == operator.

File: Breeze/src/Breeze.Wallet/Models/WalletBuildTransactionModel.cs [hunk 2]
```diff
@@ -16,13 +16,10 @@ public class WalletBuildTransactionModel
 		public Money Fee { get; set; }
 
 		[JsonProperty(PropertyName = "feePercentOfSent")]
-		public double FeePercentOfSent { get; set; }
+		public decimal FeePercentOfSent { get; set; }
 
 		[JsonProperty(PropertyName = "hex")]
 		public string Hex { get; set; }
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Models/WalletBuildTransactionModel.cs]: Where did the Transaction go?

File: Breeze/src/Breeze.Wallet/Models/WalletHistoryModel.cs [hunk 3]
```diff
@@ -13,15 +14,15 @@ public class WalletHistoryModel
...
 		public Money Amount { get; set; }
 
 		[JsonProperty(PropertyName = "confirmed")]
-		public Money Confirmed { get; set; }
+		public bool Confirmed { get; set; }
 
 		[JsonProperty(PropertyName = "timestamp")]
-		public string Timestamp { get; set; }
+		public DateTimeOffset Timestamp { get; set; }
 	}
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Models/WalletHistoryModel.cs]: The timestamp will be formatted in a pre-agreed string format before being send to the front end.

File: Breeze/src/Breeze.Wallet/Wrappers/WalletWrapper.cs [hunk 4]
```diff
@@ -84,6 +85,28 @@ public WalletModel Recover(string password, string folderPath, string name, stri
...
+			if (trimmed.Equals("alice", StringComparison.OrdinalIgnoreCase))
+				return _aliceAccount;
+			if(trimmed.Equals("bob", StringComparison.OrdinalIgnoreCase))
+				return _bobAccount;
+			throw new ArgumentException("Wrong account");
+		}
+
+		private FeeType GetFeeType(string feeType)
+		{
+			var trimmed = feeType.Trim();
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Wrappers/WalletWrapper.cs]: i think using an enum would be appropriate here

Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Wrappers/WalletWrapper.cs]: Switch + Trim().ToLowerOrdinal() please

File: Breeze/src/Breeze.Wallet/Controllers/WalletController.cs [hunk 5]
```diff
@@ -62,13 +62,12 @@ public IActionResult Load([FromBody]WalletLoadRequest walletLoad)
...
-                var wallet = this.walletWrapper.Load(walletLoad.Password, walletLoad.FolderPath, walletLoad.Name);
-                return this.Json(wallet);
-
+                this.walletWrapper.Load(walletLoad.Password, walletLoad.FolderPath, walletLoad.Name);
+                return this.Json(new SuccessModel {Success = true});
             }            
             catch (FileNotFoundException e)
             {
-				return ErrorHelpers.BuildErrorResponse(HttpStatusCode.Conflict, "This wallet already exists.", e.ToString());
+				return ErrorHelpers.BuildErrorResponse(HttpStatusCode.NotFound, "Wallet not found.", e.ToString());
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Controllers/WalletController.cs]: good spot!

File: Breeze/src/Breeze.Wallet/Controllers/WalletController.cs [hunk 6]
```diff
@@ -94,24 +93,24 @@ public IActionResult Recover([FromBody]WalletRecoveryRequest walletRecovery)
...
-                return this.Json(wallet);
+                this.walletWrapper.Recover(walletRecovery.Password, walletRecovery.FolderPath, walletRecovery.Name, walletRecovery.Network, walletRecovery.Mnemonic);
+	            return this.Json(new SuccessModel {Success = true});
 
             }
             catch (FileNotFoundException e)
             {
 				// indicates that this wallet does not exist
-				return ErrorHelpers.BuildErrorResponse(HttpStatusCode.NotFound, "Wallet not found.", e.ToString());				
+				return ErrorHelpers.BuildErrorResponse(HttpStatusCode.Conflict, "This wallet already exists.", e.ToString());
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Controllers/WalletController.cs]: I guess if a wallet will be generated we'll never get a FileNotFoundException , is that right?

File: Breeze/src/Breeze.Wallet/Controllers/WalletController.cs [hunk 7]
```diff
@@ -94,24 +93,24 @@ public IActionResult Recover([FromBody]WalletRecoveryRequest walletRecovery)
...
+				return ErrorHelpers.BuildErrorResponse(HttpStatusCode.Conflict, "This wallet already exists.", e.ToString());	
             }
             catch (Exception e)
             {
 				return ErrorHelpers.BuildErrorResponse(HttpStatusCode.BadRequest, e.Message, e.ToString());
 			}
         }
 
-	    [Route("info")]
+	    [Route("general")]
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Controllers/WalletController.cs]: "general" doesn't mean much. 'info' is a bit clearer. 'general-info' is even better.

File: Breeze/src/Breeze.Wallet/Controllers/WalletController.cs [hunk 8]
```diff
@@ -131,9 +130,53 @@ public IActionResult GetInfo([FromQuery] WalletName model)
 			}
 		}
 
+		[Route("sensitive")]
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Controllers/WalletController.cs]: decided not to include this at this point as I'm not sure it's needed by the front-end.

File: Breeze/src/Breeze.Wallet/Models/WalletInfoModel.cs [hunk 9]
```diff
@@ -6,10 +6,10 @@
 
 namespace Breeze.Wallet.Models
 {
-	public class WalletInfoModel
+	public class WalletGeneralInfoModel
 	{
-		[JsonProperty(PropertyName = "filePath")]
-		public string FilePath { get; set; }
+		[JsonProperty(PropertyName = "walletFilePath")]
+		public string WalletFilePath { get; set; }
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Models/WalletInfoModel.cs]: wallet is redundant in this case, I don't really see the point of this change.

File: Breeze/src/Breeze.Wallet/Wrappers/WalletWrapper.cs [hunk 10]
```diff
@@ -72,25 +66,24 @@ public WalletModel Load(string password, string folderPath, string name)
...
+			if (trimmed.Equals("alice", StringComparison.OrdinalIgnoreCase)
+				|| trimmed.Equals("account1", StringComparison.OrdinalIgnoreCase)
+				|| trimmed.Equals("walletaccount1", StringComparison.OrdinalIgnoreCase)
+				|| trimmed.Equals("safeaccount1", StringComparison.OrdinalIgnoreCase))
 				return _aliceAccount;
-			if(trimmed.Equals("bob", StringComparison.OrdinalIgnoreCase))
+			if(trimmed.Equals("bob", StringComparison.OrdinalIgnoreCase)
+				|| trimmed.Equals("account2", StringComparison.OrdinalIgnoreCase)
+				|| trimmed.Equals("walletaccount2", StringComparison.OrdinalIgnoreCase)
+				|| trimmed.Equals("safeaccount3", StringComparison.OrdinalIgnoreCase))
```
Comment (bokobza) [inline Breeze/src/Breeze.Wallet/Wrappers/WalletWrapper.cs]: I have no idea what this is.

**Comment by bokobza** [comment]
Thank you for this PR. 
I have a few general comments (as well as some inline):

1. Private fields shouldn't have an underscore but they should start with a lower case, as in the other classes of the project.
I didn't put the specifications in place yet but I wasn't expecting anyone to submit a pull request on these classes.

2. All the WalletJob methods: startAsync, SendTransactionAsync etc.. I don't think we're there yet. We need to have a wallet that can receive blocks and transactions and work without running its own block puller, mempool and connecting to nodes. The full node is already doing that. Hbitcoin can still do all of these things but the wallet and the walletJob need to be separated.

3. Who are Alice and Bob?

4. We agreed that Load should be a POST.

5.  success = true and success = false in Json response is redundant as we have HTTP error codes for that. This is one of the things I want to change in the APi document. I guess we don't need a SuccessModel class.

6. I'm a bit surprised to have received a pull request on this code, I didn't know anyone else was working on the endpoints. 

7. This PR shouldn't be called a refactoring. Refactoring are typically agreed in advance. They respond to a need to improve the internal workings of a solution and don't add functionalities. Please only submit pull requests on work agreed in advance to 'add' functionalities rather than to 'fix' someone else's code.

**Comment by nopara73** [comment]
1. Done. https://github.com/stratisproject/Breeze/pull/21/commits/02e2d75be563981e2ebf459f3457f75667b7a043  
2. Comment on this later...  
3. The two default account of the safe.  
	The TumbleBit whitepaper was based on these actors.  
	The TumbleBit proof of concepts (C++, Python) also hardcoded these actors.  
	NTumbleBit (C#) also hardcoded these actors, so it's pretty much official terminology now.  
4. "We agreed that Load should be a POST."  
	Sorry I misremembered, because, DevOtion had a comment on slack, something like "is the load already GET?". So I assumed, anyway changed it.  
	Fixed it: https://github.com/stratisproject/Breeze/pull/21/commits/374ae2da4330dcc00352f85a228ca29b29838de1  
5. Comment on this later...  
6. (^-^)  
7. We are at the beginning, the code should move fast without much administrative overhead. Anyway I am just exploring the project and write code where it needed. I'd like to keep this organic process for now, if that's ok for you.

File: Breeze/src/Breeze.Wallet/Controllers/WalletController.cs (hunk 6 above)
Comment (nopara73) [inline Breeze/src/Breeze.Wallet/Controllers/WalletController.cs]: Yes. This is what you intended, you just mixed up the recover and load error check:(https://github.com/stratisproject/Breeze/pull/21#discussion_r110545985)

File: Breeze/src/Breeze.Wallet/Models/WalletInfoModel.cs (hunk 9 above)
Comment (nopara73) [inline Breeze/src/Breeze.Wallet/Models/WalletInfoModel.cs]: You can have file paths of wallet transactions, monitored scriptpubkeys, etc.

**Comment by nopara73** [comment]
I dived into the code and continued the implementation, but I'm afraid I'm not able to go through and implement the API without changing things those might not be compatible with your vision for the API. I am also not a fan of to procrastinating on every little detail at this stage.  So I close this PR, but leave it up for reference to you.

**Pull request closed without merging**